pkg/runs/runsDownload.go

```diff
@@ -29,24 +30,84 @@ func DownloadArtifacts(
 
 	var err error = nil
 	var runs []galasaapi.Run
-	var artifactPaths []string
 
 	runs, err = GetRunsFromRestApi(runName, 0, 0, timeService, apiServerUrl)
 	if err == nil {
-		for _, run := range runs {
-			runId := run.GetRunId()
-			artifactPaths, err = GetArtifactPathsFromRestApi(runId, apiServerUrl)
-			if err == nil {
-				for _, artifactPath := range artifactPaths {
-					var artifactData io.Reader
-					artifactData, err = GetFileFromRestApi(runId, strings.TrimPrefix(artifactPath, "/"), apiServerUrl)
-					if err == nil {
-						err = WriteArtifactToFileSystem(fileSystem, runName, artifactPath, artifactData, forceDownload)
+		if len(runs) > 1 {
+			// get list of runs that are reRuns - get list of runs that are reRuns of each other
+			// create a map of lists of reRuns - key is queued time, value is the run
+
+			reRunsByQueuedTime := createMapOfReRuns(runs)
+
+			for _, reRunsList := range reRunsByQueuedTime {
+				if err == nil {
+					for reRunIndex, reRun := range reRunsList {
+						if err == nil {
+							directoryName := nameArtifactDownloadDirectory(reRun, reRunIndex)
+							err = downloadArtifactsToDirectory(apiServerUrl, directoryName, reRun, fileSystem, forceDownload)
+						}
 					}
 				}
 			}
+
+		} else {
+			for _, run := range runs {
+				if err == nil {
+					err = downloadArtifactsToDirectory(apiServerUrl, runName, run, fileSystem, forceDownload)
+				}
+			}
+		}
+	}
+	return err
+}
+
+func createMapOfReRuns(runs []galasaapi.Run) map[string][]galasaapi.Run {
```

Comment:
What is the map of ? What's the key ?

Reply:
eg: getRunsListpByQueuedTime ?

Doesn't this cause the queued times to be out of order, as you are putting things into a map of queued times ?

Perhaps a list of ascending queued times, with each containing a list of runs ?

Reply:
key = queued time
value = []runs
so then we can see a list of runs with the the same queued time.
this is helpful if we had two unrelated tests with the same runname which both had reruns (very edge-case)

Reply:
the queued-time doesn't need to be in order. if the runs come back sorted on queued-time and there are re-runs then these will all have the same queued-time. then these re-runs should be sorted on start-time. the earliest start-time comes in and gets added the the list associated with the queued-time in the map. then the second earliest.. to the latest so the list of runs in the map will still be ordered?